File: Zeze/Transaction/GlobalAgent.cs
﻿using System;
using System.Collections.Generic;
using System.Text;
using Zeze.Net;
using Zeze.Services;
using System.Threading.Tasks;
using NLog;
using Zeze.Services.GlobalCacheManager;

namespace Zeze.Transaction
{
    internal interface IGlobalAgent : IDisposable
    {
        // (ResultCode, State, GlobalSerialId)
        public Task<(long, int, long)> Acquire(Zeze.Beans.GlobalCacheManagerWithRaft.GlobalTableKey gkey, int state);
        public int GetGlobalCacheManagerHashIndex(Zeze.Beans.GlobalCacheManagerWithRaft.GlobalTableKey gkey);
    }

    public sealed class GlobalAgent : IGlobalAgent
    {
        internal static readonly NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();
        public GlobalClient Client { get; private set; } // 未启用cache-sync时为空。

        public void Dispose()
        {
            Stop();
        }

        public class Agent
        {
            public bool ActiveClose { get; private set; } = false;
            public Connector Connector { get; private set; }
            public Zeze.Util.AtomicLong LoginTimes { get; } = new();
            public int GlobalCacheManagerHashIndex { get; }

            private readonly Zeze.Util.AtomicLong LastErrorTime = new();
            public const long FastErrorPeriod = 10 * 1000; // 10 seconds

            public Agent(GlobalClient client, string host, int port, int globalCacheManagerHashIndex)
            {
                GlobalCacheManagerHashIndex = globalCacheManagerHashIndex;
                this.Connector = new Connector(host, port, true)
                {
                    UserState = this
                };
                client.Config.AddConnector(this.Connector);
            }

            private static void ThrowException(string msg, Exception cause = null)
            {
                var txn = Transaction.Current;
                if (txn != null)
                    txn.ThrowAbort(msg, cause);
                throw new Exception(msg, cause);
            }

            public async Task<AsyncSocket> ConnectAsync()
            {
                var so = Connector.TryGetReadySocket();
                if (null != so)
                    return so;

                if (global::Zeze.Util.Time.NowUnixMillis - LastErrorTime.Get() < FastErrorPeriod)
                    ThrowException("GlobalAgent.Connect: In Forbid Login Period");

                try
                {
                    return await Connector.GetReadySocketAsync();
                }
                catch (Exception e)
                {
                    // 并发的等待，简单用个规则：在间隔期内不再设置。
                    long now = global::Zeze.Util.Time.NowUnixMillis;
                    lock (this)
                    {
                        if (now - LastErrorTime.Get() > FastErrorPeriod)
                            LastErrorTime.GetAndSet(now);
                    }
                    ThrowException("GlobalAgent.Connect: Login Timeout", e);
                }
                return null; // never got here
            }

            public void Close()
            {
                lock (this)
                {
                    // 简单保护一下，Close 正常程序退出的时候才调用这个，应该不用保护。
                    if (ActiveClose)
                        return;
                    ActiveClose = true;
                }

                try
                {
                    var ready = Connector.TryGetReadySocket();
                    if (null != ready)
                        new NormalClose().SendForWait(ready).Task.Wait();
                }
                finally
                {
                    Connector.Stop();
                }
            }

            public void OnSocketClose(GlobalClient client, Exception ex)
            {
                lock (this)
                {
                    if (ActiveClose)
                    {
                        // active close
                        return;
                    }
                }

                GlobalAgent.logger.Warn(ex, $"GlobalAgent({Connector}) Passive Close.");

                if (Connector.IsHandshakeDone)
                {
                    foreach (var database in client.Zeze.Databases.Values)
                    {
                        foreach (var table in database.Tables)
                        {
                            table.ReduceInvalidAllLocalOnly(GlobalCacheManagerHashIndex);
                        }
                    }
                    client.Zeze.CheckpointRun();
                }
            }
        }

        internal Agent[] Agents;

        public int GetGlobalCacheManagerHashIndex(Zeze.Beans.GlobalCacheManagerWithRaft.GlobalTableKey gkey)
        {
            return gkey.GetHashCode() % Agents.Length;
        }

        public async Task<(long, int, long)> Acquire(Zeze.Beans.GlobalCacheManagerWithRaft.GlobalTableKey gkey, int state)
        {
            if (null != Client)
            {
                var agent = Agents[GetGlobalCacheManagerHashIndex(gkey)]; // hash
                var socket = await agent.ConnectAsync();

                // 请求处理错误抛出异常（比如网络或者GlobalCacheManager已经不存在了），打断外面的事务。
                // 一个请求异常不关闭连接，尝试继续工作。
                var rpc = new Acquire(gkey, state);
                await rpc.SendForWait(socket, 12000).Task;
                /*
                if (rpc.ResultCode != 0) // 这个用来跟踪调试，正常流程使用Result.State检查结果。
                {
                    logger.Warn("Acquire ResultCode={0} {1}", rpc.ResultCode, rpc.Result);
                }
                */
                if (rpc.ResultCode < 0)
                {
                    Transaction.Current.ThrowAbort("GlobalAgent.Acquire Failed");
                    // never got here
                }
                switch (rpc.ResultCode)
                {
                    case GlobalCacheManagerServer.AcquireModifyFailed:
                    case GlobalCacheManagerServer.AcquireShareFailed:
                        Transaction.Current.ThrowAbort("GlobalAgent.Acquire Failed");
                        // never got here
                        break;
                }
                return (rpc.ResultCode, rpc.Result.State, rpc.Result.GlobalSerialId);
            }
            logger.Debug("Acquire local ++++++");
            return (0, state, 0);
        }

        public async Task<long> ProcessReduceRequest(Zeze.Net.Protocol p)
        {
            var rpc = (Reduce)p;
            switch (rpc.Argument.State)
            {
                case GlobalCacheManagerServer.StateInvalid:
                    {
                        var table = Zeze.GetTable(rpc.Argument.GlobalTableKey.TableName);
                        if (table == null)
                        {
                            logger.Warn($"ReduceInvalid Table Not Found={rpc.Argument.GlobalTableKey.TableName},ServerId={Zeze.Config.ServerId}");
                            // 本地没有找到表格看作成功。
                            rpc.Result.GlobalTableKey = rpc.Argument.GlobalTableKey;
                            rpc.Result.State = GlobalCacheManagerServer.StateInvalid;
                            rpc.SendResultCode(0);
                            return 0;
                        }
                        return await table.ReduceInvalid(rpc);
                    }

                case GlobalCacheManagerServer.StateShare:
                    {
                        var table = Zeze.GetTable(rpc.Argument.GlobalTableKey.TableName);
                        if (table == null)
                        {
                            logger.Warn($"ReduceShare Table Not Found={rpc.Argument.GlobalTableKey.TableName},ServerId={Zeze.Config.ServerId}");
                            // 本地没有找到表格看作成功。
                            rpc.Result.GlobalTableKey = rpc.Argument.GlobalTableKey;
                            rpc.Result.State = GlobalCacheManagerServer.StateInvalid;
                            rpc.SendResultCode(0);
                            return 0;
                        }
                        return await table.ReduceShare(rpc);
                    }

                default:
                    rpc.Result = rpc.Argument;
                    rpc.SendResultCode(GlobalCacheManagerServer.ReduceErrorState);
                    return 0;
            }
        }

        public Zeze.Application Zeze { get;  }

        public GlobalAgent(Zeze.Application app)
        {
            Zeze = app;
        }

        public void Start(string[] hostNameOrAddress, int port)
        {
            lock (this)
            {
                if (null != Client)
                    return;

                Client = new GlobalClient(this, Zeze);
                // Raft Need. Zeze-App 自动启用持久化的全局唯一的Rpc.SessionId生成器。
                //Client.SessionIdGenerator = Zeze.ServiceManagerAgent.GetAutoKey(Client.Name).Next;

                Client.AddFactoryHandle(new Reduce().TypeId, new Service.ProtocolFactoryHandle()
                {
                    Factory = () => new Reduce(),
                    Handle = ProcessReduceRequest,
                    TransactionLevel = TransactionLevel.None
                });
                Client.AddFactoryHandle(new Acquire().TypeId, new Service.ProtocolFactoryHandle()
                {
                    Factory = () => new Acquire(),
                    TransactionLevel = TransactionLevel.None
                    // 同步方式调用，不需要设置Handle: Response Timeout 
                });
                Client.AddFactoryHandle(new Login().TypeId, new Service.ProtocolFactoryHandle()
                {
                    Factory = ()=>new Login(),
                    TransactionLevel = TransactionLevel.None
                });
                Client.AddFactoryHandle(new ReLogin().TypeId, new Service.ProtocolFactoryHandle()
                {
                    Factory = () => new ReLogin(),
                    TransactionLevel = TransactionLevel.None
                });
                Client.AddFactoryHandle(new NormalClose().TypeId, new Service.ProtocolFactoryHandle()
                {
                    Factory = () => new NormalClose(),
                    TransactionLevel = TransactionLevel.None
                });

                Agents = new Agent[hostNameOrAddress.Length];
                for (int i = 0; i < hostNameOrAddress.Length; ++i)
                {
                    var hp = hostNameOrAddress[i].Split(':');
                    if (hp.Length > 1)
                        Agents[i] = new Agent(Client, hp[0], int.Parse(hp[1]), i);
                    else
                        Agents[i] = new Agent(Client, hp[0], port, i);
                }
                Client.Start();
                foreach (var agent in Agents)
                {
                    _ = agent.ConnectAsync(); // 异步启动
                }
            }
        }

        public void Stop()
        {
            lock (this)
            {
                if (null == Client)
                    return;

                foreach (var agent in Agents)
                {
                    agent.Close();
                }
                Client.Stop();
            }
        }
    }

    public sealed class GlobalClient : Zeze.Net.Service
    {
        public GlobalClient(GlobalAgent agent, Application zeze)
            : base($"{agent.Zeze.SolutionName}.GlobalClient", zeze)
        {
        }

        public override void OnHandshakeDone(AsyncSocket so)
        {
            var agent = so.UserState as GlobalAgent.Agent;
            if (agent.LoginTimes.Get() > 0)
            {
                var relogin = new ReLogin();
                relogin.Argument.ServerId = Zeze.Config.ServerId;
                relogin.Argument.GlobalCacheManagerHashIndex = agent.GlobalCacheManagerHashIndex;
#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
                relogin.Send(so,
                    async (_) =>
                    {
                        if (relogin.IsTimeout)
                        {
                            so.Close(new Exception("GlobalAgent.ReLogin Timeout"));
                        }
                        else if (relogin.ResultCode != 0)
                        {
                            so.Close(new Exception("GlobalAgent.ReLogin Fail code=" + relogin.ResultCode));
                        }
                        else
                        {
                            agent.LoginTimes.IncrementAndGet();
                            base.OnHandshakeDone(so);
                        }
                        return 0;
                    });
#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
            }
            else
            {
                var login = new Login();
                login.Argument.ServerId = Zeze.Config.ServerId;
                login.Argument.GlobalCacheManagerHashIndex = agent.GlobalCacheManagerHashIndex;
#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
                login.Send(so,
                    async (_) =>
                    {
                        if (login.IsTimeout)
                        {
                            so.Close(new Exception("GlobalAgent.Login Timeout"));
                        }
                        else if (login.ResultCode != 0)
                        {
                            so.Close(new Exception($"GlobalAgent.Logoin Error {login.ResultCode}"));
                        }
                        else
                        {
                            agent.LoginTimes.IncrementAndGet();
                            base.OnHandshakeDone(so);
                        }
                        return 0;
                    });
#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
            }
        }

        public override void DispatchProtocol(Protocol p, ProtocolFactoryHandle factoryHandle)
        {
            _ = Util.Mission.CallAsync(factoryHandle.Handle, p);
        }

        public override void OnSocketClose(AsyncSocket so, Exception e)
        {
            base.OnSocketClose(so, e);
            var agent = so.UserState as GlobalAgent.Agent;
            if (null == e)
                e = new Exception("Peer Normal Close.");
            agent.OnSocketClose(this, e);
        }
    }
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/GlobalAgent.java
package Zeze.Transaction;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.atomic.AtomicLong;
import Zeze.Application;
import Zeze.Beans.GlobalCacheManagerWithRaft.GlobalTableKey;
import Zeze.Net.AsyncSocket;
import Zeze.Net.Connector;
import Zeze.Net.Service;
import Zeze.Services.GlobalCacheManager.Acquire;
import Zeze.Services.GlobalCacheManager.Login;
import Zeze.Services.GlobalCacheManager.NormalClose;
import Zeze.Services.GlobalCacheManager.ReLogin;
import Zeze.Services.GlobalCacheManager.Reduce;
import Zeze.Services.GlobalCacheManagerServer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public final class GlobalAgent implements IGlobalAgent {
	private static final Logger logger = LogManager.getLogger(GlobalAgent.class);

	public static final class Agent {
		private static final long FastErrorPeriod = 10 * 1000; // 10 seconds

		private final Connector connector;
		private final AtomicLong LoginTimes = new AtomicLong();
		private final int GlobalCacheManagerHashIndex;
		private boolean ActiveClose;
		private volatile long LastErrorTime;

		public Agent(GlobalClient client, String host, int port, int _GlobalCacheManagerHashIndex) {
			connector = new Zeze.Net.Connector(host, port, true);
			connector.UserState = this;
			GlobalCacheManagerHashIndex = _GlobalCacheManagerHashIndex;
			client.getConfig().AddConnector(connector);
		}

		public AtomicLong getLoginTimes() {
			return LoginTimes;
		}

		public int getGlobalCacheManagerHashIndex() {
			return GlobalCacheManagerHashIndex;
		}

		private void ThrowException(String msg, Throwable cause) {
			var txn = Transaction.getCurrent();
			if (txn != null)
				txn.ThrowAbort(msg, cause);
			throw new RuntimeException(msg, cause);
		}

		public AsyncSocket Connect() {
			try {
				var so = connector.TryGetReadySocket();
				if (so != null)
					return so;

				synchronized (this) {
					if (System.currentTimeMillis() - LastErrorTime < FastErrorPeriod)
						ThrowException("GlobalAgent In FastErrorPeriod", null); // abort
					// else continue
				}

				return connector.WaitReady();
			} catch (Throwable abort) {
				var now = System.currentTimeMillis();
				synchronized (this) {
					if (now - LastErrorTime > FastErrorPeriod)
						LastErrorTime = now;
				}
				ThrowException("GlobalAgent Login Failed", abort);
			}
			return null; // never got here.
		}

		public void Close() {
			try {
				synchronized (this) {
					// 简单保护一下重复主动调用 Close
					if (ActiveClose)
						return;
					ActiveClose = true;
				}
				var ready = connector.TryGetReadySocket();
				if (ready != null)
					new NormalClose().SendForWait(ready).Wait();
			} finally {
				connector.Stop(); // 正常关闭，先设置这个，以后 OnSocketClose 的时候判断做不同的处理。
			}
		}

		public void OnSocketClose(GlobalClient client, Throwable ignoredEx) {
			synchronized (this) {
				if (ActiveClose)
					return; // Connector 的状态在它自己的回调里面处理。
			}

			if (connector.isHandshakeDone()) {
				for (var database : client.getZeze().getDatabases().values()) {
					for (var table : database.getTables())
						table.ReduceInvalidAllLocalOnly(getGlobalCacheManagerHashIndex());
				}
				client.getZeze().CheckpointRun();
			}
		}
	}

	private final Application Zeze;
	private GlobalClient Client;
	public Agent[] Agents;

	public GlobalAgent(Application app) {
		Zeze = app;
	}

	public Application getZeze() {
		return Zeze;
	}

	public GlobalClient getClient() {
		return Client;
	}

	@Override
	public int GetGlobalCacheManagerHashIndex(GlobalTableKey gkey) {
		return gkey.hashCode() % Agents.length;
	}

	@Override
	public void close() {
		try {
			Stop();
		} catch (Throwable e) {
			throw new RuntimeException(e);
		}
	}

	@Override
	public AcquireResult Acquire(GlobalTableKey gkey, int state) {
		if (Client != null) {
			var agent = Agents[GetGlobalCacheManagerHashIndex(gkey)]; // hash
			var socket = agent.Connect();

			// 请求处理错误抛出异常（比如网络或者GlobalCacheManager已经不存在了），打断外面的事务。
			// 一个请求异常不关闭连接，尝试继续工作。
			var rpc = new Acquire(gkey, state);
			try {
				rpc.SendForWait(socket, 12000).get();
			} catch (InterruptedException | ExecutionException e) {
				var trans = Transaction.getCurrent();
				if (trans == null)
					throw new GoBackZeze("Acquire", e);
				trans.ThrowAbort("Acquire", e);
				// never got here
			}
			/*
			if (rpc.ResultCode != 0) // 这个用来跟踪调试，正常流程使用Result.State检查结果。
			{
			    logger.Warn("Acquire ResultCode={0} {1}", rpc.ResultCode, rpc.Result);
			}
			*/
			if (rpc.getResultCode() == GlobalCacheManagerServer.AcquireModifyFailed
					|| rpc.getResultCode() == GlobalCacheManagerServer.AcquireShareFailed) {
				var trans = Transaction.getCurrent();
				if (trans == null)
					throw new GoBackZeze("GlobalAgent.Acquire Failed");
				trans.ThrowAbort("GlobalAgent.Acquire Failed", null);
				// never got here
			}
			return new AcquireResult(rpc.getResultCode(), rpc.Result.State, rpc.Result.GlobalSerialId);
		}
		logger.debug("Acquire local ++++++");
		return new AcquireResult(0, state, 0);
	}

	public int ProcessReduceRequest(Reduce rpc) {
		switch (rpc.Argument.State) {
		case GlobalCacheManagerServer.StateInvalid:
			var table1 = Zeze.GetTable(rpc.Argument.GlobalTableKey.getTableName());
			if (null == table1) {
				logger.warn("ReduceInvalid Table Not Found={},ServerId={}",
						rpc.Argument.GlobalTableKey.getTableName(), Zeze.getConfig().getServerId());
				// 本地没有找到表格看作成功。
				rpc.Result.GlobalTableKey = rpc.Argument.GlobalTableKey;
				rpc.Result.State = GlobalCacheManagerServer.StateInvalid;
				rpc.SendResultCode(0);
				return 0;
			}
			return table1.ReduceInvalid(rpc);

		case GlobalCacheManagerServer.StateShare:
			var table2 = Zeze.GetTable(rpc.Argument.GlobalTableKey.getTableName());
			if (table2 == null) {
				logger.warn("ReduceShare Table Not Found={},ServerId={}",
						rpc.Argument.GlobalTableKey.getTableName(), Zeze.getConfig().getServerId());
				// 本地没有找到表格看作成功。
				rpc.Result.GlobalTableKey = rpc.Argument.GlobalTableKey;
				rpc.Result.State = GlobalCacheManagerServer.StateInvalid;
				rpc.SendResultCode(0);
				return 0;
			}
			return table2.ReduceShare(rpc);

		default:
			rpc.Result = rpc.Argument;
			rpc.SendResultCode(GlobalCacheManagerServer.ReduceErrorState);
			return 0;
		}
	}

	public synchronized void Start(String[] hostNameOrAddress, int port) throws Throwable {
		if (Client != null)
			return;

		Client = new GlobalClient(this, Zeze);
		Client.AddFactoryHandle(new Reduce().getTypeId(), new Service.ProtocolFactoryHandle<>(
				() -> new Reduce(), this::ProcessReduceRequest, TransactionLevel.None));

		Client.AddFactoryHandle(new Acquire().getTypeId(), new Service.ProtocolFactoryHandle<>(
				() -> new Acquire(), null, TransactionLevel.None));
		Client.AddFactoryHandle(new Login().getTypeId(), new Service.ProtocolFactoryHandle<>(
				() -> new Login(), null, TransactionLevel.None));
		Client.AddFactoryHandle(new ReLogin().getTypeId(), new Service.ProtocolFactoryHandle<>(
				() -> new ReLogin(), null, TransactionLevel.None));
		Client.AddFactoryHandle(new NormalClose().getTypeId(), new Service.ProtocolFactoryHandle<>(
				() -> new NormalClose(), null, TransactionLevel.None));

		Agents = new Agent[hostNameOrAddress.length];
		for (int i = 0; i < hostNameOrAddress.length; i++) {
			var hp = hostNameOrAddress[i].split(":", -1);
			Agents[i] = new Agent(Client, hp[0], hp.length > 1 ? Integer.parseInt(hp[1]) : port, i);
		}

		Client.Start();

		for (var agent : Agents) {
			try {
				agent.Connect();
			} catch (Throwable ex) {
				// 允许部分GlobalCacheManager连接错误时，继续启动程序，虽然后续相关事务都会失败。
				logger.error("GlobalAgent.Connect", ex);
			}
		}
	}

	public synchronized void Stop() throws Throwable {
		if (Client == null)
			return;
		for (var agent : Agents)
			agent.Close();
		Client.Stop();
		Client = null;
	}
}